### ISSUE TITLE
Module 'LayoutTestBase' not found

### ISSUE THREAD

**Issue by cconeil** [linkedin/LayoutTest-iOS#8]
I'm getting an error when trying to build for tests, even without `#import "LayoutTest.h"` anywhere.

<img width="807" alt="screen shot 2016-02-11 at 11 42 12 am" src="https://cloud.githubusercontent.com/assets/1817861/12988265/bb7db2b6-d0b4-11e5-9664-11129e9331a7.png">

It seems like the linker is having a hard time with the `@import`. I'm only linking these pods with my test target, which has a minimum build of iOS 7.0. I'm using cocoapods `0.38.2`. I'm facing this issue in `pod 'LayoutTest', '1.1.0'` and `pod 'LayoutTest', '~> 1.1.1'`

Is there anything I can do fix this issue? I would love to be able to integrate these view tests! It seems like a very well done testing framework. Thanks!

**Comment by chentoo** [comment]
me too ~

**Comment by liamjdouglas** [comment]
Don't think iOS 7 support will be possible as Embedded Frameworks aren't supported. Podspec will need updated to reflect this. @plivesey can hopefully confirm though, not sure what minimum target linkedin are supporting.

**Comment by plivesey** [comment]
Hmmmm. I'd like to support iOS 7, because there's nothing in the code which is iOS 8 specific.

I think this was needed for swift support because using ```#import "LayoutTestBase.h"``` didn't work with Allow non-modular Includes.

Does changing to ```#import "LayoutTestBase.h"``` fix the problem for you guys?

Before abandoning iOS 7, I'd go for:

- Adding iOS 8 as the minimum build target for LayoutTest/Swift. This is because Swift isn't supported for cocoapods < iOS 8 (maybe we don't need this because cocoapods will already warn people?)
- Finding a work around to make this work on iOS 7. My first idea is to do something like ``` #if BUILD_SESSINGS.target == 'iOS 7'```

I think I'll have some time to look at this today.

**Comment by plivesey** [comment]
Ok. I've investigated, and the problem seems to be around use_frameworks!

I have two projects. One is Objective-C and has this Podfile:

```
target :'TestLayoutObjcTests' do
pod 'LayoutTest'
end
```

It has a deployment target of iOS 7 and fails to compile. If I change ```@import LayoutTestBase``` to ```#import "LYTLayoutPropertyTester.h"```, it compiles. Also, if I add ```use_frameworks!``` to the start of the podfile, it compiles. Therefore, I think the bug occurs when not using ```use_frameworks!```.

In my other project, I have a swift project which has the following Podfile:

```
use_frameworks!

target :'TestLayoutPodTests' do
pod 'LayoutTest/Swift'
end
```

This will compile normally. If I change ```@import LayoutTestBase``` to ```#import "LYTLayoutPropertyTester.h"```, it fails to compile. Therefore, this fix will not work because it will break existing projects.

So, iOS 7 is not actually the problem. It's an absence of use_frameworks. However, it also seems like this problem must have been hit before since it will happen whenever you have a pod which depends on another pod. I'll do some research and see what I can find. Anyone else have any ideas on how to fix?

**Comment by plivesey** [comment]
In the meantime, if you need a workaround, you could change your Podfile to ```use_frameworks!```, but completely understand if you don't want to do this or this isn't possible (is this even supported for iOS 7? I know it's not if you use swift...not sure about objective-c).

**Comment by plivesey** [comment]
I've opened a stack overflow question: http://stackoverflow.com/questions/35468498/include-of-non-modular-header-inside-framework-error-how-do-you-make-transitiv
Had no luck investigating this this morning.

**Comment by cconeil** [comment]
Hi @plivesey - I just want to give you some feedback that I've been following your investigation and really appreciate you looking into this! I'm going to try a couple of things tomorrow and I'll report back here.

I originally tried adding `use_frameworks!` for only my testing target, but I didn't have any luck. Right now, it doesn't look all of our dependencies play nice with `use_frameworks` (maybe because of iOS 7?), so hopefully we can find a different solution.

Anyway, I'll keep you updated on any progress I make. Thanks again!

**Comment by plivesey** [comment]
Great. Thanks @cconeil.
I have another temporary work around. Change your podfile to:

```
pod 'LayoutTestBase'
```

Then, copy and paste LYTLayoutTestCase into your project. Change the top of the file manually to ```#import "LYTLayoutPropertyTester.h"```.

Now, you should be able to write layout tests. This solution is clearly not acceptable long term, but a possible workaround for now.

**Comment by plivesey** [comment]
Oh, and also, we may have some progress...someone has been experimenting with custom module maps. Will update the thread later today.

**Comment by plivesey** [comment]
Some progress is being made: https://github.com/goyalankit/LayoutTestSample/commit/f53a5d8bab7e4d6eb1a545caa0a6b7b3b5c2f90c

**Comment by plivesey** [comment]
Also: https://guides.cocoapods.org/syntax/podspec.html#module_map

**Comment by plivesey** [comment]
Note: the parse SDK seems to get around this problem somehow:
https://github.com/ParsePlatform/Parse-SDK-iOS-OSX

See this file which is a .h which imports something from a dependency:
https://github.com/ParsePlatform/Parse-SDK-iOS-OSX/blob/master/Parse/PFCloud.h

I can't work out what they are doing differently though yet.

**Comment by plivesey** [comment]
This looks promising. Something like:
```
#if __has_include(<RealmMapView/RealmMapView-BridgingHeader.h>)
@import RBQFetchedResultsController;
#else
#import <RBQFetchedResultsController/RBQFetchRequest.h>
#endif
```
https://github.com/CocoaPods/CocoaPods/issues/4420

**Comment by plivesey** [comment]
According to this thread, this issue may have been fixed in Cocoapods-1.0.0-beta. Looks like it was committed Nov 12th, but 39.0 was released in October.

**Comment by plivesey** [comment]
This seems to work in my local tests:
```
#if __has_include(<LayoutTestBase/LayoutTestBase-BridgingHeader.h>)
@import LayoutTestBase;
#else
#import <LayoutTestBase/LYTLayoutPropertyTester.h>
#endif
```

**Comment by plivesey** [comment]
I've pushed some commits to the repo which I think fix the problem.

@cconeil @chentoo could you please try your projects against HEAD to make sure it works for you? I haven't actually released the version, so you'll have to do:

```
pod 'LayoutTest', :head
```

I think that's the right syntax?

**Comment by plivesey** [comment]
If it works for your projects, I'll do a release tomorrow.

**Comment by cconeil** [comment]
Hey @plivesey - it looks like it's linking properly now! Thanks so much for your work.

With that being said, I'm still running into some issues. In order to get it work, I took your advice and used `pod 'LayoutTest', :head` (I also had to add `pod 'LayoutTestBase', :head` to get it to compile).

Here are my issues:

1. I'm not able to `#import <LayoutTestBase/LayoutTestBase.h>`. I was looking at the github project, and it looks like that file exists, but I can't seem to find it after doing a `pod install`. Maybe there is a problem with your `.podspec`? I've solved this issue by importing each file from `LayoutTestBase` on its own.
2. There might be some sort of infinite loop? Since I've updated to the `:head` of both of these pods, my one test takes minutes to run, and finally stops with this error:
 ```*** Test operation was canceled. If you believe this error represents a bug, please attach the log file at /var/folders/dh/rm4hrgm50y92q0lr77x_q4cm0000gn/T/com.apple.dt.XCTest-status/Session-2016-02-19_13:43:12-rbFvd9.log```. 
That didn't look right to me.

Thanks for all of your help! Let me know if you want me to try anything else.

**Comment by plivesey** [comment]
Thanks for checking it out. Yeah, makes sense that you need both LayoutTest and LayoutTestBase. As for the imports, where is this code? Is this in your app code?

I'm not sure if there is a 'real' import all file. I should probably add one...so used to the @import syntax :).

Can you paste the code of your test? I can give some advice on how to debug then.

Thanks for your help getting this working.

**Comment by plivesey** [comment]
Btw, if you checkout the project, in TestProjects/SampleAppiOS7 there's an example of a very simple test which passes. So something works, but potentially there's a bug elsewhere.

**Comment by plivesey** [comment]
@cconeil any update? I'd like to release this soon :), but if it's not working for you...maybe it won't help much?

**Comment by cconeil** [comment]
@plivesey sorry about that - I was able to get it working with `:head` so you should feel free to go ahead and release.

Thanks!

**Comment by plivesey** [comment]
I'll do it today - thanks.

**Comment by plivesey** [comment]
Going to close this ticket since we have a working solution. Fix will be in 1.2.0.
